### ISSUE TITLE
Speed signal of the ego vehicle

### ISSUE THREAD

**Issue by ashwinsarvesh** [lgsvl/simulator#856]
Hello, 

I would like to get the speed of the ego vehicle from the simulation and send it to a camera that has a GPS receiver (2.5mm jack). Is this possible to get the speed of the simulated vehicle? 

The reason why I need this: This camera has ADAS functions like lane departure and forward collision warning. For these features to work, I need this speed of the simulated ego vehicle. I place this camera in front of the simulator.

**Comment by EricBoiseLGSVL** [comment]
@ashwinsarvesh You can enable the sensor visualizations and have any value you need.  If you need a different UI implementation, you will need to edit simulator code or make a sensor that displays UI data.  It could even be as simple as an OnGUI().

![SpeedHelp](https://user-images.githubusercontent.com/45106230/87489150-1b124e80-c5f7-11ea-925e-b70a0127bd45.png)

**Comment by lemketron** [comment]
You could also monitor (on ROS, ROS2, or CyberRT) the CAN Bus messages (including speed).

For example, in /apollo/canbus/chassis you will see the speed in meters per second labeled as "speed_mps". You could monitor this data programmatically and then determine how to present it to your ADAS camera.

You mentioned a GPS antenna, but generating fake GPS radio signals to trick a stationary ADAS camera into thinking it is moving is non-trivial, and not something the simulator will be able to help you with. If you can figure out how to inject the speed into the ADAS camera (or whatever it is connected to) that would be much easier.

**Comment by ashwinsarvesh** [comment]
Hello,

Thank you for your reply.
I need to extract the following signals of the ego vehicle from the Simulator/Apollo to connect it to the ADAS camera.
- The vehicle speed,
- The vehicle ignition signal (12/24V),
- The Vehicle CAN bus (High/Low),
- The vehicle left turn signal and right turn signal.
Is it possible to extract this information from the ego vehicle to give it to the camera?

The attached picture contains the wiring of the camera.
![Wiring](https://user-images.githubusercontent.com/67329705/88078226-7d7dc880-cb7c-11ea-89b7-e0e55a76200c.jpg)

**Comment by rongguodong** [comment]
On the real car, these signals will be passed to your camera driver via those wires.
When you use our simulator, you may implement a bridge plugin (please refer to our ROS/ROS2/Cyber bridges). So the bridge will receive all messages from our simulator, and send the useful ones to your camera driver.

**Comment by ashwinsarvesh** [comment]
Thank you for your reply
Are these data which I listed above given by the simulator or Apollo?

Regarding your idea, Unfortunately, this camera does not have a ROS driver. :/  So I guess this idea will not work. 
What do you say?

**Comment by rongguodong** [comment]
I believe we have all those information available in the simulator (may need more information on that ignition signal -- what does it mean? Is it just an on/off status?)
Even if the camera dose not have a ROS driver, it must have some driver, right? You can build a bridge plugin working with that driver.

**Comment by ashwinsarvesh** [comment]
ok, I will check on the driver part.
Ignition is a (12/24V) signal. When the vehicle ignition switch is turned to ACC (Accessory) position, this 12/24V is passed to this signal which powers on the camera.

**Comment by ashwinsarvesh** [comment]
> I believe we have all those information available in the simulator (may need more information on that ignition signal -- what does it mean? Is it just an on/off status?)
> Even if the camera dose not have a ROS driver, it must have some driver, right? You can build a bridge plugin working with that driver.

How do I access this data from the simulator?

**Comment by rongguodong** [comment]
Please read the code to understand how the current sensors access data and send out messages.

**Comment by ashwinsarvesh** [comment]
Yes, I will do that.
But can you tell me whether I have to read lgsvl code or Apollo code to know how to access data and send out messages?
If possible, a specific part to read.

**Comment by EricBoiseLGSVL** [comment]
Look at sensors and bridge code to see how we send data.  They are in their respective folders in Assets/Scripts/Sensors or /Bridge

**Comment by ashwinsarvesh** [comment]
ok got it. Thanks
CanBusData is converted to Apollo specific Apollo.ChassisMsg type
Is it being published as a ROS message, it should be, right? Otherwise, how can it be sent over the ROS bridge.
Am I right?

**Comment by ashwinsarvesh** [comment]
Hello,

I am planning to use [socketcan_bridge](http://wiki.ros.org/socketcan_bridge?distro=noetic) to send CAN messages received on the ROS topic /apollo/canbus/chassis to the SocketCAN device which is a CAN hardware connected to a PC.

This socket_bridge converts can messages received on a ROS topic to can messages and sends out to hardware.
What I want to know is, in the topic /apollo/canbus/chassis are can the messages received as ROS topic?

**Comment by EricBoiseLGSVL** [comment]
@ashwinsarvesh We don't support Socketcan bridge currently but I think you can create a custom bridge to subscribe to the messages.  Right @hadiTab ?

**Comment by ashwinsarvesh** [comment]
@EricBoiseLGSVL Can you elaborate what you mean by not support Socketcan_bridge?
I thought I could install socketcan_bridge and then use it to send the "can_msgs/Frame" messages from the simulator (after converting from CanBusData type). I mean, I thought both were independent and Socketcan_bridge doesn't have to do anything with the simulator.

And by a custom bridge, you mean custom bridge plugin?

**Comment by EricBoiseLGSVL** [comment]
@ashwinsarvesh Yes we support ROS, ROS2 and CyberRT.  This new bridge type will need to be added or create a custom solution.

**Comment by ashwinsarvesh** [comment]
@EricBoiseLGSVL I get it now. ROS bridges and Cyber RT bridges are adapted to be supported by LGSVL but socketcan_bridge is not adapted. Is my understanding correct?

**Comment by EricBoiseLGSVL** [comment]
@hadiTab is the expert on this but yes you are correct

**Comment by ashwinsarvesh** [comment]
Can you tell me what I have to refer to implement a custom bridge plugin?

**Comment by EricBoiseLGSVL** [comment]
Sorry, this feature has not released yet but it should be soon.  In the meantime you can look to how we implemented the other bridge types and start to create your own in source.

**Comment by ashwinsarvesh** [comment]
Thank you for your reply
I'll surely check this method
In this method,looks like socketcan_bridge can work independently without
the need of LGSVL support. Am I right?

**Comment by ashwinsarvesh** [comment]
Thank you for your reply.
I'll check this method
Where can I find the below details for the individual signals(for eg. Speed, left turn signal, right turn signal)
As the "can_msgs/Frame" asks for the below details. It would be helpful if I can find these.
![image](https://user-images.githubusercontent.com/67329705/88785568-a5c78180-d191-11ea-9a92-0c00ca24b798.png)

**Comment by ashwinsarvesh** [comment]
Hello,

I am sorry if I was not clear with my previous question.
For each can signal produced by the simulator(eg. speed, left and right turn signal), 

- where can I find It's message information like CAN identifiers and so on. 
- Is the data sent out in bytes?

I need this info to convert to the can_msgs/Frame type. As of now, I can just see the can bus data [here](https://github.com/lgsvl/simulator/blob/master/Assets/Scripts/Bridge/Data/CanBusData.cs)
